ipie/addons/thermal/qmc/thermal_afqmc.py

```diff
@@ -0,0 +1,472 @@
+                            self.params.num_walkers,
+                            self.params.num_steps_per_block,
+                            self.mpi_handler,
+                            self.params.pop_control_method,
+                            verbose=self.verbose)
+        
+        self.get_env_info()
+        self.setup_estimators(estimator_filename, additional_estimators=additional_estimators)
+        
+        synchronize()
+        comm = self.mpi_handler.comm
+        self.tsetup += time.time() - ft_setup
+
+        # Propagate.
+        total_steps = self.params.num_steps_per_block * self.params.num_blocks
+        # TODO: This magic value of 2 is pretty much never controlled on input.
+        # Moreover I'm not convinced having a two stage shift update actually
+        # matters at all.
+        neqlb_steps = 2.0 / self.params.timestep
+        nslices = numpy.rint(self.params.beta / self.params.timestep).astype(int)
+
+        for step in range(1, total_steps + 1):
+            synchronize()
+            start_path = time.time()
+
+            for t in range(nslices):
+                if self.verbosity >= 2 and comm.rank == 0:
+                    print(" # Timeslice %d of %d." % (t, nslices))
+
+                start = time.time()
+                self.propagator.propagate_walkers(
+                        self.walkers, self.hamiltonian, self.trial, eshift, debug=self.debug)
+
+                self.tprop_fbias = self.propagator.timer.tfbias
+                self.tprop_update = self.propagator.timer.tupdate
+                self.tprop_vhs = self.propagator.timer.tvhs
+                self.tprop_gemm = self.propagator.timer.tgemm
+                
+                start_clip = time.time()
+                if t > 0:
+                    wbound = self.pcontrol.total_weight * 0.10
+                    xp.clip(self.walkers.weight, a_min=-wbound, a_max=wbound,
+                            out=self.walkers.weight)  # In-place clipping.
+
+                synchronize()
+                self.tprop_clip += time.time() - start_clip
+
+                start_barrier = time.time()
+                if t % self.params.pop_control_freq == 0:
+                    comm.Barrier()
+
+                self.tprop_barrier += time.time() - start_barrier
+                self.tprop += time.time() - start
+                
+                if (t > 0) and (t % self.params.pop_control_freq == 0):
+                    start = time.time()
+                    self.pcontrol.pop_control(self.walkers, comm)
+                    synchronize()
+                    self.tpopc += time.time() - start
+                    self.tpopc_send = self.pcontrol.timer.send_time
+                    self.tpopc_recv = self.pcontrol.timer.recv_time
+                    self.tpopc_comm = self.pcontrol.timer.communication_time
+                    self.tpopc_non_comm = self.pcontrol.timer.non_communication_time
+                
+                # Print cut.
```

Comment:
Remove commented out code. Do you intend to make it a configurable option to print out time slice information?

Reply:
Sure--if you think it's a good idea?

Reply:
I think it could be a useful diagnostic in the future. You don't need to do it now. Just open an issue to track.

Reply:
Added a keyword argument to `ThermalAFQMC.run`